src/batch.rs

```diff
@@ -27,61 +24,130 @@ fn gen_128_bits<R: RngCore + CryptoRng>(mut rng: R) -> [u64; 4] {
+        tup: (
+            VerificationKeyBytes<SpendAuth>,
+            Signature<SpendAuth>,
+            Scalar,
+        ),
+    ) -> Self {
+        let (vk_bytes, sig, c) = tup;
+        Inner::SpendAuth { vk_bytes, sig, c }
+    }
+}
+
+impl From<(VerificationKeyBytes<Binding>, Signature<Binding>, Scalar)> for Inner {
+    fn from(tup: (VerificationKeyBytes<Binding>, Signature<Binding>, Scalar)) -> Self {
+        let (vk_bytes, sig, c) = tup;
+        Inner::Binding { vk_bytes, sig, c }
+    }
+}
+
 /// A batch verification item.
 ///
 /// This struct exists to allow batch processing to be decoupled from the
 /// lifetime of the message. This is useful when using the batch verification API
 /// in an async context.
-pub struct Item<T: SigType> {
-    vk_bytes: VerificationKeyBytes<T>,
-    sig: Signature<T>,
-    c: Scalar,
+pub struct Item {
+    inner: Inner,
 }
 
-impl<'msg, M: AsRef<[u8]>, T: SigType + ?Sized>
-    From<(VerificationKeyBytes<T>, Signature<T>, &'msg M)> for Item<T>
+impl<'msg, M: AsRef<[u8]>>
+    From<(
+        VerificationKeyBytes<SpendAuth>,
+        Signature<SpendAuth>,
+        &'msg M,
+    )> for Item
 {
-    fn from(tup: (VerificationKeyBytes<T>, Signature<T>, &'msg M)) -> Self {
+    fn from(
+        tup: (
+            VerificationKeyBytes<SpendAuth>,
+            Signature<SpendAuth>,
+            &'msg M,
+        ),
+    ) -> Self {
         let (vk_bytes, sig, msg) = tup;
         // Compute c now to avoid dependency on the msg lifetime.
         let c = HStar::default()
             .update(&sig.r_bytes[..])
             .update(&vk_bytes.bytes[..])
             .update(msg)
             .finalize();
-        Self { vk_bytes, sig, c }
+        Self {
+            inner: Inner::SpendAuth { vk_bytes, sig, c },
+        }
     }
 }
 
+impl<'msg, M: AsRef<[u8]>> From<(VerificationKeyBytes<Binding>, Signature<Binding>, &'msg M)>
+    for Item
+{
+    fn from(tup: (VerificationKeyBytes<Binding>, Signature<Binding>, &'msg M)) -> Self {
+        let (vk_bytes, sig, msg) = tup;
+        // Compute c now to avoid dependency on the msg lifetime.
+        let c = HStar::default()
+            .update(&sig.r_bytes[..])
+            .update(&vk_bytes.bytes[..])
+            .update(msg)
+            .finalize();
+        Self {
+            inner: Inner::Binding { vk_bytes, sig, c },
+        }
+    }
+}
+
+// This would ideally be an associated type with `Verifier` but
+// generic associated types are unstable:
+// https://github.com/rust-lang/rust/issues/44265
+type BatchTuple<T: SigType> = (VerificationKeyBytes<T>, Signature<T>, Scalar);
+
 /// A batch verification context.
-pub struct Verifier<T: SigType> {
+pub struct Verifier {
     /// Signature data queued for verification.
-    signatures: Vec<(VerificationKeyBytes<T>, Scalar, Signature<T>)>,
+    signatures: Vec<Item>,
     /// Caching this count avoids a hash traversal to figure out
     /// how much to preallocate.
     batch_size: usize,
```

Comment:
I think this can be eliminated now as `batch_size = signatures.len()`; in the Ed25519 code, this was used because the signatures are stored in a tree to save common terms, but here there's just a flat list.